### PR TITLE
Rename input -> address for address genes in hidden/outp…

### LINKED ISSUE #242: Use "address gene" for genes encoding indices of input nodes
Currently we use "input gene" for these which gets very confusing because "inputs" are also the nodes that hold variable leaf values. As discussed with @HenrikMettler we should rather use "address gene" for these, as this also agrees with the nomenclature in the literature.

### PR DESCRIPTION
**Pull request by HenrikMettler**
- renamed input_gene to address gene to describe the nodes from which hidden/output nodes read. Consistent with general literature. Closes #242.

### CODE REVIEW

File: cgp/node_impl.py [hunk 1]
```diff
@@ -11,35 +11,35 @@ class ConstantFloat(OperatorNode):
 
 
 class Add(OperatorNode):
-    """A node that adds its two inputs."""
+    """A node that adds its two addresses."""
```
Comment (jakobj) [inline cgp/node_impl.py:14]: mhm. nope, i don't think we should rename this here ;) i mean we don't need to distort language, the important thing is to use good variables names in the code.

File: examples/example_parametrized_nodes.py [hunk 2]
```diff
@@ -5,7 +5,7 @@
 Example demonstrating the use of Cartesian genetic programming for a
 regression task that requires fine tuning of constants in parametrized
 nodes. This is achieved by introducing a new node, "ParametrizedAdd"
-which produces a scaled and shifted version of the sum of its inputs.
+which produces a scaled and shifted version of the sum of its addresses.
```
Comment (jakobj) [inline examples/example_parametrized_nodes.py:8]: (see above, i would keep input here, as this is easily understandable, but what is "the sum of its addresses"?)

File: test/test_cartesian_graph.py [hunk 3]
```diff
@@ -20,7 +20,7 @@ def test_direct_input_output():
     )
     genome.randomize(np.random)
 
-    genome[-2:] = [0, ID_NON_CODING_GENE]  # set inputs for output node to input node
+    genome[-2:] = [0, ID_NON_CODING_GENE]  # set address node for output node to input node
```
Comment (jakobj) [inline test/test_cartesian_graph.py:23]:
```suggestion
    genome[-2:] = [0, ID_NON_CODING_GENE]  # set address for output node to input node
```

File: cgp/cartesian_graph.py [hunk 4]
```diff
@@ -181,7 +181,7 @@ def _determine_active_nodes(self) -> Dict[int, Set[Node]]:
                 continue
             active_nodes_by_hidden_column_idx[self._hidden_column_idx(node.idx)].add(node)
 
-            for i in node.input_nodes:
+            for i in node.address_nodes:
```
Comment (jakobj) [inline cgp/cartesian_graph.py:184]: (see below, maybe better `addresses`?)

File: examples/example_parametrized_nodes.py (hunk 2 above)
Comment (jakobj) [reply examples/example_parametrized_nodes.py:8]: (also in code below)

File: test/test_genome.py [hunk 5]
```diff
@@ -359,15 +359,15 @@ def test_mutation_rate(rng_seed, mutation_rate):
     genome.randomize(rng)
 
     def count_n_immutable_genes(n_inputs, n_outputs, n_rows):
-        length_per_region = genome.primitives.max_arity + 1  # function gene + input gene addresses
+        length_per_region = genome.primitives.max_arity + 1  # function gene + gene addresses
```
Comment (jakobj) [inline test/test_genome.py:362]:
```suggestion
        length_per_region = genome.primitives.max_arity + 1  # function gene + address gene
```

File: test/test_genome.py [hunk 6]
```diff
@@ -447,7 +447,7 @@ def test_only_silent_mutations(genome_params, mutation_rate, rng_seed):
     genome.dna = dna_fixed
     graph = CartesianGraph(genome)
     active_regions = graph.determine_active_regions()
-    length_per_region = genome.primitives.max_arity + 1  # function gene + input gene addresses
+    length_per_region = genome.primitives.max_arity + 1  # function gene + gene addresses
```
Comment (jakobj) [inline test/test_genome.py:450]:
```suggestion
    length_per_region = genome.primitives.max_arity + 1  # function gene + address gene
```

File: test/test_genome.py [hunk 7]
```diff
@@ -488,10 +488,10 @@ def test_permissible_values(genome_params):
...
 
     permissible_function_gene_values = np.arange(len(genome._primitives._primitives))
-    permissible_inputs_first_internal_column = np.array([0, 1])
-    permissible_inputs_second_internal_column = np.array([0, 1, 2, 3, 4])
-    permissible_inputs_third_internal_column = np.array([0, 1, 2, 3, 4, 5, 6, 7])
-    permissible_inputs_output_gene = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10])
+    permissible_addresses_first_internal_column = np.array([0, 1])
+    permissible_addresses_second_internal_column = np.array([0, 1, 2, 3, 4])
+    permissible_addresses_third_internal_column = np.array([0, 1, 2, 3, 4, 5, 6, 7])
+    permissible_addresses_output_gene = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10])
```
Comment (jakobj) [inline test/test_genome.py:494]:
```suggestion
    permissible_addresses_output = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10])
```

File: test/test_node.py [hunk 8]
```diff
@@ -5,23 +5,23 @@
...
-    provided, Node.inputs only returns the initial <arity> inputs,
+def test_addresses_are_cut_to_match_arity():
+    """Test that even if a list of addresses longer than the node arity is
+    provided, Node.addresses only returns the initial <arity> addresses,
     ignoring the inactive genes.
 
     """
     idx = 0
-    input_nodes = [1, 2, 3, 4]
+    address_nodes = [1, 2, 3, 4]
```
Comment (jakobj) [inline test/test_node.py:15]:
```suggestion
    addresses = [1, 2, 3, 4]
```

**Review by jakobj — CHANGES_REQUESTED** [review]
great, thanks for working on this @HenrikMettler, a tedious task.

i've added some inline comments that should be addresses to improve the usage of "address" a bit.

File: cgp/node.py [hunk 9]
```diff
@@ -33,25 +33,25 @@ class Node:
...
         """Init function.
 
         Parameters
         ----------
         idx : int
             id of the node.
-        inputs : List[int]
-            List of integers specifying the input nodes to this node.
+        addresses : List[int]
+            List of integers specifying the address nodes to this node.
```
Comment (jakobj) [inline cgp/node.py:49]:
```suggestion
            List of integers specifying the address of input nodes to this node.
```

File: cgp/node.py [hunk 10]
```diff
@@ -175,7 +175,7 @@ class OperatorNode(Node):
 
     def __init_subclass__(cls: Type["OperatorNode"]) -> None:
         super().__init_subclass__()
-        OperatorNode._extract_input_names_from_def_output(cls)
+        OperatorNode._extract_address_names_from_def_output(cls)
```
Comment (jakobj) [inline cgp/node.py:178]: i think this should still be called "input names", "address names" is not very intuitive is it?

File: test/test_node.py [hunk 11]
```diff
@@ -536,22 +536,22 @@ class CustomAdd(cgp.OperatorNode):
 
 def test_repr():
     idx = 0
-    input_nodes = [1, 2, 3, 4]
+    address_nodes = [1, 2, 3, 4]
```
Comment (jakobj) [inline test/test_node.py:539]:
```suggestion
    addresses = [1, 2, 3, 4]
```

File: cgp/node.py (hunk 10 above)
Comment (HenrikMettler) [reply cgp/node.py:178]: agree

File: cgp/node.py [hunk 12]
```diff
@@ -76,15 +76,15 @@ def idx(self) -> int:
     def __repr__(self) -> str:
         return (
             f"{self.__class__.__name__}(idx: {self.idx}, active: {self._active}, "
-            f"arity: {self._arity}, input_nodes {self._input_nodes}"
+            f"arity: {self._arity}, address_nodes {self._addresses}"
```
Comment (jakobj) [inline cgp/node.py:79]:
```suggestion
            f"arity: {self._arity}, addresses {self._addresses}"
```

Comment (jakobj) [reply cgp/node.py:79]: or "addresses of inputs", this should easily understandable as its printed to the screen

**Review by jakobj — APPROVED** [review]
great stuff! :+1: 

one more small comment, then please squash the commits

**Comment by jakobj** [comment]
excellent :+1: and merging

**Pull request merged**